src/Identity/Utilities/ServiceCollectionExtensions.cs

```diff
@@ -45,11 +47,33 @@ public static class ServiceCollectionExtensions
             .AddCustomTokenRequestValidator<CustomTokenRequestValidator>()
             .AddProfileService<ProfileService>()
             .AddResourceOwnerValidator<ResourceOwnerPasswordValidator>()
-            .AddPersistedGrantStore<PersistedGrantStore>()
             .AddClientStore<ClientStore>()
             .AddIdentityServerCertificate(env, globalSettings)
             .AddExtensionGrantValidator<WebAuthnGrantValidator>();
 
+        if (CoreHelpers.SettingHasValue(globalSettings.Grants.RedisConnectionString))
+        {
+            // If we have redis, prefer it
+
+            // Add the original persisted grant store via it's implementation type
+            // so we can inject it right after.
+            services.AddSingleton<PersistedGrantStore>();
+
+            services.AddSingleton<IPersistedGrantStore>(sp =>
+            {
+                return new RedisPersistedGrantStore(
+                    ConnectionMultiplexer.Connect(globalSettings.Grants.RedisConnectionString),
+                    sp.GetRequiredService<ILogger<RedisPersistedGrantStore>>(),
+                    sp.GetRequiredService<PersistedGrantStore>() // Fallback grant store
```

Comment:
Why not do this in the ctor?

Reply:
Specifically the `ConnectionMultiplexer.Connect`? Because if I were to write unit tests for this I could instantiate `RedisPersistedGrantStore` with a mocked version of `IConnectionMultiplexer`. But I don't have such tests and I don't think I would write tests that mock redis behavior vs just running it against redis locally so if you want me to move it there I absolutely can.

Reply:
This just looks dirty instead of letting DI do its thing. What if we create and register a new IGrantConnectionMultiplexer or something and register that?

Reply:
I can do that but in my opinion the real fix here is to use the new [keyed services](https://andrewlock.net/exploring-the-dotnet-8-preview-keyed-services-dependency-injection-support/) in .NET 8. We can name a specific instance of `IConnectionMultiplexer` to be used for `"grants"`. Then we don't need a special marker interface.

Reply:
I agree. I don't care either way if we feel it might be better to just wait for that.